### DISCUSSION TITLE
source IP as slave id on modbus TCP instead of uid

### DISCUSSION THREAD

**Discussion by ickedude** [pymodbus-dev/pymodbus#1905] [General]
Hi,

I am trying to start an TCP server with more than one slave context.
On one slave I am able to configure the unit identifier, but another one looks like it follows modbus messaging implementation guide and therefor it sends messages with unit identifier 255.
I cannot create a slave to the server context with id 255, because the [server context validates that the slave id is between 1 and 247](https://github.com/pymodbus-dev/pymodbus/blob/61fac339d1db214026003f74593250be3a908073/pymodbus/datastore/context.py#L161).
I could overwrite the validation, but what happens if I have a second slave following this implementation? I cannot distinguish between them.

As stated below, the uid is useless on TCP. So I would suggest  to use the source IP as slave id, but I am somehow unable to understand the socket frame to replace uid in the header with the source IP and circumvent all validations.
_Can anyone help? Should I create an issue for that?_

From the modbus messaging implementation guide:

> Unit Identifier
This field is used for routing purpose when addressing a device on a MODBUS+
or MODBUS serial line sub-network. In that case, the “Unit Identifier” carries the
MODBUS slave address of the remote device:
[...]
**On TCP/IP, the MODBUS server is addressed using its IP address; therefore, the
MODBUS Unit Identifier is useless.** The value 0xFF has to be used.
When addressing a MODBUS server connected directly to a TCP/IP network,
it’s recommended not using a significant MODBUS slave address in the “Unit
Identifier” field.
[...]
0xFF is recommended for the “Unit Identifier" as non-
significant value.
Remark : The value 0 is also accepted to communicate directly to a
MODBUS/TCP device.

**Comment by janiversen** [comment]
Our server currently only allows one unit id, this is due to a global structure. It should ideally be for each slave id.

Feel free to to raise an issue, but it is unlikely to be solved anytime soon, so much better if you submit a PR.

**Reply by ickedude** [reply]
Is it possible to pin point me to a place where I can get the source IP of the connection. Maybe I can work on a PR.

**Reply by janiversen** [reply]
it's a command line parameter.

**Comment by ickedude** [comment]
For testing purpose I modified [execute in async_io.py](https://github.com/pymodbus-dev/pymodbus/blob/61fac339d1db214026003f74593250be3a908073/pymodbus/server/async_io.py#L180C9-L181C55) to:
```diff
         if self.server.request_tracer:
-            self.server.request_tracer(request, *addr)
+            self.server.request_tracer(request, *addr, peer=self.transport.get_extra_info('peername'))
```
and added a request_tracer
```python
def test(request, *addr, **kwargs):
    pprint(kwargs)

master = await StartAsyncTcpServer(
    context=server_context,
    identity=identity,
    address=('192.168.10.10', 5020),
    framer=ModbusSocketFramer,
    request_tracer=test,
)
```
The output now shows the peer IP:
```
DEBUG:pymodbus.logging:Connected to server
[...]
2023-12-07 09:13:55,146 DEBUG logging:102 Factory Request[WriteMultipleRegistersRequest': 16]
DEBUG:pymodbus.logging:Factory Request[WriteMultipleRegistersRequest': 16]
{'peer': ('192.168.10.22', 33272)}
```
At the moment I think the easiest and smallest change would be to modify the request and replace the uid with the peer IP and overwrite the [validation in `__set_item__` of ModbusServerContext](https://github.com/pymodbus-dev/pymodbus/blob/61fac339d1db214026003f74593250be3a908073/pymodbus/datastore/context.py#L161C9-L161C34). I will try to implement it in the next days.
Appreciate comments on that idea.

**Reply by janiversen** [reply]
It looks like an idea, I cannot see if it has any sideeffect. You cannot set a tracer in the production code, only for testing !
